girder/models/model_base.py

```diff
@@ -772,10 +770,30 @@ def _setAccess(self, doc, id, entity, level, save, flags=None, user=None, force=
                 'flags': flags
             }
             entry['flags'] = self._validateFlags(doc, user, entity, entry, force)
-            doc['access'][entity].append(entry)
+            # becuase we're iterating this operation is not necesarily atomic
+            for index, perm in enumerate(doc['access'][entity]):
+                if perm['id'] == id:
+                    # if the id already exists we want to update with a $set
+                    doc['access'][entity][index] = entry
+                    update['$set'] = {'%s.%s' % (key, index): entry}
+                    break
+            else:
+                doc['access'][entity].append(entry)
+                update['$push'] = {key: entry}
+        # set remove query
+        else:
+            update['$pull'] = {key: {'id': id}}
+            for perm in doc['access'][entity]:
+                if perm['id'] == id:
+                    doc['access'][entity].remove(perm)
 
         if save:
-            doc = self.save(doc)
+            if '_id' not in doc:
+                doc = self.save(doc)
+            else:
+                doc = self.collection.find_one_and_update(
+                    {'_id': ObjectId(doc['_id'])}, update,
+                    return_document=pymongo.ReturnDocument.AFTER)
```

Comment:
This breaks a subtle part of this function's contract, which is that prior to this change, calling with `save=True` could be used to update other fields of the document besides just `access`.

To fix this, inside this block we could iterate over all the top-level keys of this dictionary *besides* "access", and add them to the `$update` list.